src/backend/distributed/connection/remote_commands.c

```diff
@@ -531,3 +555,238 @@ GetRemoteCommandResult(MultiConnection *connection, bool raiseInterrupts)
 
 	return result;
 }
+
+
+/*
+ * CreateMultiConnectionWaiter creates a MultiConnectionWaiter for the given list
+ * of connections. Connections should not be in a closed state.
+ */
+MultiConnectionWaiter *
+CreateMultiConnectionWaiter(List *connectionList)
+{
+	MultiConnectionWaiter *waiter = NULL;
+	ListCell *connectionCell = NULL;
+	int totalConnectionCount = list_length(connectionList);
+
+	waiter = (MultiConnectionWaiter *) palloc0(sizeof(MultiConnectionWaiter));
+	waiter->connectionCount = 0;
+	waiter->currentOffset = 0;
+
+	waiter->connections =
+		(MultiConnection **) palloc(totalConnectionCount * sizeof(MultiConnection *));
+
+	foreach(connectionCell, connectionList)
+	{
+		MultiConnection *connection = (MultiConnection *) lfirst(connectionCell);
+
+		waiter->connections[waiter->connectionCount] = connection;
+		waiter->connectionCount++;
+	}
+
+	waiter->events = (WaitEvent *) palloc0(totalConnectionCount * sizeof(WaitEvent));
+
+	return waiter;
+}
+
+
+/*
+ * AwaitResultsOnConnections waits for I/O events and returns a list of connections
+ * that which finished or failed. Each connection is returned at most once across
+ * multiple calls.
+ *
+ * If all connections have been returned, AwaitResultsOnConnections returns an
+ * empty list.
+ *
+ * To keep track of which connections were already returned, pending connections
+ * are moved to the end of the connections array, starting at currentOffset.
+ */
+List *
+AwaitResultsOnConnections(MultiConnectionWaiter *waiter, bool raiseInterrupts)
+{
+	MultiConnection **allConnections = waiter->connections;
+	bool connectionReady[waiter->connectionCount];
+	bool foundReadyConnections = false;
+	List *readyConnectionList = NIL;
+	WaitEventSet *waitEventSet = NULL;
+	int connectionCount = waiter->connectionCount;
+	int pendingConnectionsStartIndex = waiter->currentOffset;
+	int pendingConnectionCount = connectionCount - pendingConnectionsStartIndex;
+	int connectionIndex = 0;
+	bool cancelled = false;
+
+	if (waiter->currentOffset == connectionCount)
+	{
+		/* already processed all connections */
+		return NIL;
+	}
+
+	/*
+	 * Ideally, we'd store the WaitEventSet in MultiConnectionWaiter and reuse it across
+	 * different calls to AwaitResultsOnConnections. However, we currently don't have a
+	 * way to stop listening for a connection in a WaitEventSet once it's ready, since
+	 * ModifyWaitEvent has an assert preventing the caller from clearing the event mask
+	 * event masks. This should be revisited in future postgres versions.
+	 */
+	waitEventSet = CreateWaitEventSet(CurrentMemoryContext, pendingConnectionCount + 2);
+
+	for (connectionIndex = pendingConnectionsStartIndex;
+		 connectionIndex < connectionCount; connectionIndex++)
+	{
+		MultiConnection *connection = allConnections[connectionIndex];
+		int socket = PQsocket(connection->pgConn);
+		int eventMask = WL_SOCKET_READABLE;
+
+		int sendStatus = PQflush(connection->pgConn);
```

Comment:
Well, that was what I was thinking: The code got simpler by letting errors fall through to `WaitEventSetWait`.

We can't easily skip over connections here and then still go into `WaitEventSetWait` since in that case the event position is no longer aligned with the connection index and we'd need some other way of getting the index out of the event. However, we could skip waiting altogether in case of an error and that would also allow checking for `PGINVALID_SOCKET`. Will change it.